### DISCUSSION TITLE
bind to specific network

### DISCUSSION THREAD

**Discussion by yair-mantis** [rustdesk/rustdesk#2286] [General]
i have an issue with one of my external interfaces firewall, blocking NAT.   
but, rustDesk, when presented with two WAN facing interfaces always chooses the blocked one. 
the internet suggested using something like [forcebindip](forcebindip)

>ForceBindIP is a freeware Windows application that will inject itself into another application and alter how certain Windows socket calls are made, allowing you to force the other application to use a specific network interface / IP address. This is useful if you are in an environment with multiple interfaces and your application has no option to bind to a specific interface

while freeware it offers no source and i dont feel comfortable injecting a dll into rustDesk.
any other suggestions?

**Comment by yair-mantis** [comment]
another (open sourced) alternative [bindip](https://github.com/katlogic/bindip)

note [this interesting segment ](https://github.com/katlogic/bindip#setting-up-default-interface) about system metrics (in windows)  
basically you're system will use a `metrics` param tied to your interface to choose default outgoing interface.

**Comment by GrennKren** [comment]
I originally wanted to use ForceBindIP, but since its last update was in 2017, it’s basically abandoned. BindIP is even older, with no updates for the past 10 years.  

While looking for alternatives, I came across this repo: [https://github.com/saeedkhatami/FBI](https://github.com/saeedkhatami/FBI)  

The problem is, whether it’s ForceBindIP or this new version, I’m not sure how to inject it into the RustDesk service.  

That’s when I took another look at BindIP and found the **"Setting up default interface"** section, which lets you manually adjust the interface metric.  

Since my only goal was to make RustDesk use a different network interface (without worrying about interface priority), I tried changing the metric, and it worked!  
[https://www.tenforums.com/tutorials/92180-change-network-adapter-connection-priorities-windows-10-a.html](https://www.tenforums.com/tutorials/92180-change-network-adapter-connection-priorities-windows-10-a.html)

**Comment by rustdesk** [comment]
Pull request is welcome.

The jobs incude:

- `upd/tcp` functions (pub listen/connect) in https://github.com/rustdesk/hbb_common/blob/main/src/socket_client.rs
- web sockets in https://github.com/rustdesk/hbb_common/blob/main/src/websocket.rs
- reqwest http request in https://github.com/rustdesk/rustdesk/blob/master/src/common.rs
- restart service when `bind_interface` option changes, add this to CheckIfRestart function in https://github.com/rustdesk/rustdesk/blob/master/src/ipc.rs
- fall back to any address if the binded interface does not exist (bind fails).
- add a droplist of network interfaces on GUI to set `bind_interface` option, `Settings` -> `Network`.

**Reply by rustdesk** [reply]
# Polished by AI

## Contributions Welcome!

Pull requests are very welcome for any of the following improvements:

### 1. Network Binding / Interface Selection Features

**Goal**: Allow users to bind RustDesk server/listener sockets to a specific network interface (instead of always binding to `0.0.0.0` / all interfaces).

Main tasks:

- **Add `bind_interface` configuration option**
  - Type: string (interface name e.g. "eth0", "wlan0", "enp3s0", "Ethernet", or empty = default behavior)
  - Default: empty string (current behavior = bind to all interfaces)

- **Implement interface binding in socket creation functions**
  Locations:
  - `hbb_common/src/socket_client.rs`
    - `tcp::listen()`
    - `tcp::connect()`  (local bind address when connecting outbound — if applicable)
    - `udp` functions if they create/bind sockets
  - `hbb_common/src/websocket.rs` — all server/client websocket binding/listen code

  → When `bind_interface` is set → resolve interface name to IP → use `.bind(ip)` instead of default

- **Fallback logic when the chosen interface does not exist or bind fails**
  - If binding to the specified interface IP fails (interface gone, no permission, etc.)
  - → automatically fall back to binding to `0.0.0.0` (all interfaces) or `::` for IPv6
  - → log a clear warning message once

### 2. Auto-restart logic when `bind_interface` changes

File: `rustdesk/src/ipc.rs`

- In `CheckIfRestart` / service restart condition checking function:
  - Add check whether the **effective** `bind_interface` value has changed
  - If changed → trigger service restart (same way as when port or other critical settings change)

### 3. GUI: Network interface dropdown

Location: Settings → Network tab (or create it if missing)

- Add a new setting: **"Bind to network interface"**
- Show a drop-down list containing:
  - "All interfaces" (empty string / default)
  - All currently available network interfaces (names + maybe IP in parentheses)
    Examples:
    - All interfaces
    - Ethernet (192.168.1.45)
    - Wi-Fi (10.0.0.123)
    - Docker bridge (172.17.0.1)
    - …

Implementation hints:
- Use `pnet` or `if-addrs` or `network-interface` crate to enumerate interfaces (choose one already allowed in dependencies or add it)
- Refresh list when opening the settings page
- Select current value from config
- Save interface **name** (not IP) to config

### Summary of files most likely needing changes

- `hbb_common/src/socket_client.rs`        → socket binding logic
- `hbb_common/src/websocket.rs`            → websocket server binding
- `rustdesk/src/common.rs`                 → reqwest/client changes if outbound interface binding is desired
- `rustdesk/src/ipc.rs`                    → restart detection
- GUI settings code (probably `rustdesk/src/ui/` or `src/client.rs` / `src/server.rs`)

Thank you for any help implementing this feature — it is very useful for multi-homed servers, VPN users, Docker setups, etc.

**Comment by tombueng** [comment]
Took a shot at the interface binding from the roadmap here. Two branches (submodule split):
- hbb_common (engine): https://github.com/tombueng/hbb_common/tree/bind-interface
- rustdesk (UI + submodule bump): https://github.com/tombueng/rustdesk/tree/bind-interface

Adds Settings → Network → Network interface: a binding mode, the address or interface, and a Strict (no fallback) toggle. Off by default.

Config keys / options:
- `bind-mode`: `` (empty = off, default) | `ip` | `interface`
- `bind-value`: source IP (for `ip` mode) or interface name (for `interface` mode)
- `bind-strict`: `Y` = fail instead of falling back to all interfaces; empty = off (default)

Binds incoming and outgoing. `ip` mode sets the source address; `interface` mode also pins the socket to the device (`SO_BINDTODEVICE` Linux/Android, `IP_BOUND_IF` macOS/iOS, `IP_UNICAST_IF` Windows) so egress follows it regardless of routing — needed to beat a full-tunnel VPN.

Tested: Linux build + unit/integration tests, Windows/macOS/iOS arms cross-compiled, `IP_UNICAST_IF` runtime-checked on Windows. macOS/iOS not runtime-tested.

Before opening PRs: are the option names and the Settings placement OK?

**Comment by 21pages** [comment]
Available in rustdesk serve pro https://github.com/rustdesk/rustdesk-server-pro/releases/tag/1.8.6

Client: https://github.com/rustdesk/rustdesk/pull/15675
